### DISCUSSION TITLE
Upload files without writing to database

### DISCUSSION THREAD

**Discussion by kokoshneta** [filamentphp/filament#6231] [Help] [answered]
I can find several old questions here about how to get Filament to write the name/path of an uploaded file to the database, to which the answer is that it does that automatically. I cannot figure out a way to **prevent** this, however.

On a product edit page, I have (or want to have) several file upload components. They all save the uploaded file in a subfolder specific to each component, with the product’s GTIN number as the file name. Some of the components (for images) should save multiple versions of the uploaded image in different subfolders.¹

There are no tables or columns in the database for these files. Everything is based purely on the presence or absence of each file in the filesystem. As such, I do not want the upload components to correspond to any database column or to write any value to the database at all.

But a name is mandatory for the component, and I haven’t been able to find a way to instruct the component _not_ to use that name as a model property and attempt to write to the database. I thought perhaps `saveRelationshipsUsing(NULL)` might do it, but an uploaded file is of course not a relationship, so it doesn’t.

Is there a way to make a file uploader that doesn’t react with the model or its database table **at all** (other than using a property to generate file names)?

---

¹  I assume I should be able to use `saveUploadedFileUsing()` for this, calling `saveUploadedFiles()` repeatedly inside the callback, but I haven’t really got to it properly yet. Preliminary tests just show what looks like a blank `dd()` page, black with a single line across it.

**Comment by danharrin** [comment, marked as answer]
You need to unset the file upload array keys from the form data before its saved: https://filamentphp.com/docs/2.x/admin/resources/editing-records#customizing-data-before-saving

**Reply by kokoshneta** [reply]
Ah, I would never have thought of that, but it makes sense! Just the matter of the tripartite saving, then, but that’s a different question.

If anyone else needs to do this, the array key to unset is the one used as the name of the component; so if your `XyzResource::form()` method has something like this:

```php
FileUpload::make('product_image')
  ->getUploadedFileNameForStorageUsing(static fn (?Model $record) => "{$record->id}.jpg")
```

Then in your `EditXyz` class, you would add this:

```php
protected function mutateFormDataBeforeSave(array $data): array {
  unset($data['product_image']);
  return $data;
}
```

**Reply by HelgeSverre** [reply]
I recently needed to do something similar to this where i used a Select component that loaded data from an external api and set various other fields on select, it would be nice if this use case would be natively supported, by calling a method on the field ex ->visualOnly() or something like that.

**Reply by danharrin** [reply]
For a Select or other non-file-upload component you could use `dehydrated(false)` on that component. I think its in the docs.

**Reply by atorscho** [reply]
I guess there is no other way of using something like `dehydrated(false)` for FileUpload, right?
